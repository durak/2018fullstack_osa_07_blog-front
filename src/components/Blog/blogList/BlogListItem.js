import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import { Card, Icon, Dimmer, Header } from 'semantic-ui-react'

export class BlogListItem extends React.Component {
  state = {}

  handleShow = () => this.setState({ active: true })
  handleHide = () => this.setState({ active: false })

  handleLike = (blog) => () => {
    const msg = `you liked ${blog.title} by ${blog.author}`
    this.props.blogLike(blog)
    this.props.notify(msg, 'message')
  }

  render() {
    const { active } = this.state
    const { blog } = this.props

    return (
      <Card>

        <Dimmer.Dimmable as={Card.Content} dimmed={active} blurring
          onMouseEnter={this.handleShow}
          onMouseLeave={this.handleHide}
        >
          <Card.Header>
            <span className="blogtitle">{blog.title}</span>
          </Card.Header>

          <Card.Meta>
            <span>
              by {blog.author}
            </span>
          </Card.Meta>

          <Card.Description>
            {blog.title}
          </Card.Description>

          <Dimmer active={active} inverted>
            <Header as="h2">
              <Link to={`blogs/${blog.id}`} > View </Link>
            </Header>
          </Dimmer>

        </Dimmer.Dimmable>

        <Card.Content extra>
          <span>
            <Icon name="like" onClick={this.handleLike(blog)} className="blogVote" />
            {blog.likes} votes
          </span>
          <span className="right floated">
            <Icon name="comments" />{blog.comments.length} comments
          </span>
        </Card.Content>

      </Card>
    )
  }
}

export default BlogListItem

BlogListItem.propTypes = {
  blog: PropTypes.object.isRequired,
  blogLike: PropTypes.func.isRequired,
  notify: PropTypes.func.isRequired
}