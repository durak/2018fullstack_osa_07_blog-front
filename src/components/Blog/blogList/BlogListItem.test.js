import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import { BlogListItem } from './BlogListItem'
import { BrowserRouter as Router } from 'react-router-dom'



describe('<BlogListItem />', () => {
  let bloglistItemComponent
  const mockHandleLike = jest.fn()
  const mockNotify = jest.fn()

  const blog = {
    id: 'ifj430foewf',
    title: 'test blog',
    author: 'test author',
    url: 'http://testurl.com',
    likes: 99,
    user: { name: 'test user full name', username: 'testusername' },
    comments: [
      { _id: 'fdsalkfjsldkjfsldkfjlksdjflskd', comment: 'some' },
      { _id: 'firoeitgrjeogifdogjidogijsdofg', comment: 'thing' }
    ]
  }


  beforeEach(() => {
    bloglistItemComponent = shallow(
      <BlogListItem
        blog={blog}
        notify={mockNotify}
        blogLike={mockHandleLike}
      />
    )
  })

  const findHostNodeWithText = (component, text) => {
    return component
      .findWhere(n => n.text().includes(text))
      .hostNodes()
  }

  it('renders correctly', () => {
    const tree = renderer.create(
      <Router>
        <BlogListItem
          blog={blog}
          notify={mockNotify}
          blogLike={mockHandleLike}
        />
        </Router>
      
    ).toJSON();
    expect(tree).toMatchSnapshot()
  })

  it('displays blog author, title', () => {
    const titleNode = bloglistItemComponent.find('.blogtitle')
    const authorNode = findHostNodeWithText(bloglistItemComponent, 'by')

    expect(titleNode.text()).toContain(blog.title)
    expect(authorNode.text()).toContain(blog.author)
    expect(authorNode.is('span')).toEqual(true)
  })

  it('clicking the like button calls like handler function', () => {
    const like = bloglistItemComponent.find('.blogVote')    
    like.at(0).simulate('click')
    expect(mockHandleLike.mock.calls.length).toBe(1)
    expect(mockNotify.mock.calls.length).toBe(1)
  })

})
