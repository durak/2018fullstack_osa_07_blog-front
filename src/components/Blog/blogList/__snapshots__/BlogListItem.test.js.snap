// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BlogListItem /> renders correctly 1`] = `
<div
  className="ui card"
  href={undefined}
  onClick={[Function]}
>
  <div
    className="content blurring dimmable"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="header"
    >
      <span
        className="blogtitle"
      >
        test blog
      </span>
    </div>
    <div
      className="meta"
    >
      <span>
        by 
        test author
      </span>
    </div>
    <div
      className="description"
    >
      test blog
    </div>
    <div
      className="ui inverted dimmer"
      onClick={[Function]}
    >
      <div
        className="content"
      >
        <h2
          className="ui header"
        >
          <a
            href="/blogs/ifj430foewf"
            onClick={[Function]}
          >
             View 
          </a>
        </h2>
      </div>
    </div>
  </div>
  <div
    className="extra content"
  >
    <span>
      <i
        aria-hidden="true"
        className="like icon blogVote"
        onClick={[Function]}
      />
      99
       votes
    </span>
    <span
      className="right floated"
    >
      <i
        aria-hidden="true"
        className="comments icon"
      />
      2
       comments
    </span>
  </div>
</div>
`;
